### PR TITLE
opc-ua module v2 for .net standard loader.

### PR DESCRIPTION
**Pull request by marcschier**

### CODE REVIEW

**Review by mregen — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1,30 +1,84 @@
...
+This reference implementation demonstrates how the Azure IoT Gateway SDK can be used to connect to existing OPC UA servers and send JSON encoded telemetry data from these servers in OPC UA "Pub/Sub" format (using a JSON payload) to Azure IoT Hub. All transport protocols supported by the Gateway SDK can be used, i.e. HTTPS, AMQP and MQTT. The transport is selected in the transport setting in gateway_config.json.
 
 This module uses the OPC Foundations's OPC UA reference stack and therefore licensing restrictions apply. Visit http://opcfoundation.github.io/UA-.NETStandardLibrary/ for OPC UA documentation and licensing terms.
 
-## Operating System Compatibility
-Since this reference implementation is written for .NET, Windows 7, 8, 8.1 & 10 are supported right now. Once the Gateway SDK supports .NET Standard, so will this module and then Linux will also be supported.
+# Azure IoT Gateway SDK compatibility
+The current version of the Proxy module is targeted at the Azure IoT Gateway SDK at commit 09bbcb7feaf5acc3913abd722b96e993238edd0c.
```
Comment (barnstee) [inline README.md:9]: Better use a release version number of the GW SDK here.

File: README.md [hunk 2]
```diff
@@ -1,30 +1,84 @@
...
 
-## Directory Structure
+Use the following command line to clone the compatible version Azure IoT Gateway SDK, then follow the build instructions included:
```
Comment (barnstee) [inline README.md:11]: Please mention that the GW needs to be built using the enable_dotnet_binding flag.

File: README.md [hunk 3]
```diff
@@ -1,30 +1,84 @@
...
+                    }
+                ]
+            },
+```
+
+# Running the module
+
+To run the module and have it publish to IoT Hub, configure the name of your Hub (JSON field ```"IoTHubName"```) and the IoT Hub device ID and shared access key to use (JSON fields ```"Id"``` and ```"SharedAccessKey"```) in your version of ```gateway_config.json```.  Note that if you use a "Mapping" module in your configuration, you can omit the ```"SharedAccessKey"``` field.  Finally, ensure that the right native module is configured, based on your platform (i.e. iothub.dll for Windows, libiothub.so for Linux, etc.).
+
+To build a sample gateway to host the OPC-UA module - along with the module itself - clone the gateway SDK repo to your device and run the build script with the ```-i <root-of-sdk>```.  The resulting release folder will contain not just the module and managed dependencies, but also a native IoT Hub proxy module as well as a ```sample_gateway``` executable that you can pass the updated JSON configuration to.
```
Comment (barnstee) [inline README.md]: Having to clone and build the GW SDK is a regression form "V1" of the GW module. If the GW SDK team cannot provide NuGet packages soon, please build them yourself (like I did) and use them here.

File: src/Opc.Ua.Client.Module/Module.cs [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+                    "Directory";
+            if (Configuration.SecurityConfiguration.RejectedCertificateStore.StorePath == null)
+                Configuration.SecurityConfiguration.RejectedCertificateStore.StorePath =
+                    "OPC Foundation/CertificateStores/RejectedCertificates";
+
+            if (Configuration.SecurityConfiguration.ApplicationCertificate == null)
+                Configuration.SecurityConfiguration.ApplicationCertificate = new CertificateIdentifier();
+            if (Configuration.SecurityConfiguration.ApplicationCertificate.StoreType == null)
+                Configuration.SecurityConfiguration.ApplicationCertificate.StoreType =
+                    "Directory";
```
Comment (barnstee) [inline src/Opc.Ua.Client.Module/Module.cs]: X509Store

File: src/Opc.Ua.Client.Module/Module.cs [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+            if (Configuration.SecurityConfiguration.ApplicationCertificate.StorePath == null)
+                Configuration.SecurityConfiguration.ApplicationCertificate.StorePath =
+                    "OPC Foundation/CertificateStores/MachineDefault";
```
Comment (barnstee) [inline src/Opc.Ua.Client.Module/Module.cs]: CurrentUser\UA_MachineDefault

File: src/Opc.Ua.Client.Module/Module.cs [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+                Module.Configuration.ApplicationName,
+                60000,
+                new UserIdentity(new AnonymousIdentityToken()),
+                null);
+
+            if (_session != null)
+            {
+                var subscription = new Subscription(_session.DefaultSubscription);
+                subscription.PublishingInterval = PublishingInterval;
+                // ...
```
Comment (barnstee) [inline src/Opc.Ua.Client.Module/Module.cs]: ?

File: src/Opc.Ua.Client.Module/Module.cs [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+                    return;
+                }
+
+                JsonEncoder encoder = new JsonEncoder(monitoredItem.Subscription.Session.MessageContext, false);
+                string hostname = monitoredItem.Subscription.Session.ConfiguredEndpoint.EndpointUrl.DnsSafeHost;
+                if (hostname == "localhost")
+                {
+                    hostname = Utils.GetHostName();
+                }
+                encoder.WriteString("HostName", hostname);
```
Comment (barnstee) [inline src/Opc.Ua.Client.Module/Module.cs]: Please port changes from Publisher to this

File: src/Opc.Ua.Client.Module/Module.cs [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+                foreach (string url in endpoint.Server.DiscoveryUrls)
+                {
+                    updatedDiscoveryUrls.Add(Utils.ReplaceLocalhost(url, discoveryUrl.DnsSafeHost));
+                }
+
+                endpoint.Server.DiscoveryUrls = updatedDiscoveryUrls;
+            }
+        }
+
+        private EndpointDescription SelectUaTcpEndpoint(EndpointDescriptionCollection endpointCollection)
```
Comment (barnstee) [inline src/Opc.Ua.Client.Module/Module.cs]: This needs to be refactored with a per session security mode selection from JSON.

**Review by barnstee — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (marcschier) [inline README.md]: It is not necessary to clone, you can build the module without it.

File: README.md (hunk 2 above)
Comment (marcschier) [inline README.md]: Will do.

File: README.md (hunk 1 above)
Comment (marcschier) [reply README.md:9]: Does not exist yet.  But agreed, this should happen.  We can wait for an official release, or we can change the readme later.

File: src/Opc.Ua.Client.Module/Module.cs (hunk 4 above)
Comment (marcschier) [reply src/Opc.Ua.Client.Module/Module.cs]: Will do

File: src/Opc.Ua.Client.Module/Module.cs (hunk 5 above)
Comment (marcschier) [reply src/Opc.Ua.Client.Module/Module.cs]: Will do

File: src/Opc.Ua.Client.Module/Module.cs (hunk 8 above)
Comment (marcschier) [reply src/Opc.Ua.Client.Module/Module.cs]: Will do

File: README.md (hunk 3 above)
Comment (marcschier) [reply README.md]: It is not necessary to clone, you can build the module without cloning the SDK.   Lets take offline.

File: README.md (hunk 2 above)
Comment (marcschier) [reply README.md:11]: Actually it needs the --enable_dotnetcore_binding flag.  Added

File: src/Opc.Ua.Client.Module/Module.cs (hunk 5 above)
Comment (marcschier) [reply src/Opc.Ua.Client.Module/Module.cs]: Shouldn't it be forward slash?  Does backslash work on Linux?

File: src/Opc.Ua.Client.Module/Module.cs (hunk 6 above)
Comment (marcschier) [reply src/Opc.Ua.Client.Module/Module.cs]: I assume the ? was about the // ... .  Adding a better comment.

File: src/Opc.Ua.Client.Module/Module.cs (hunk 5 above)
Comment (barnstee) [reply src/Opc.Ua.Client.Module/Module.cs]: yes.

File: src/Opc.Ua.Client.Module/Module.cs (hunk 6 above)
Comment (barnstee) [reply src/Opc.Ua.Client.Module/Module.cs]: correct.

**Review by barnstee — APPROVED** [review]
Leave a comment

**Pull request merged**